### ISSUE TITLE
线性回归算法

### ISSUE THREAD

**Issue by dongzhebin** [dongzhebin/NN#2]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# 设置matplotlib支持中文显示
matplotlib.rcParams['font.family'] = 'Arial Unicode MS' # 或者 'SimHei', 'Microsoft YaHei' 等你系统支持的中文字体
matplotlib.rcParams['axes.unicode_minus'] = False # 解决负号'-'显示为方块的问题

# --- 0. 定义文件路径 ---
# 请确保这里的路径指向你存放数据文件的实际位置
train_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/train.csv"
test_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/test.csv"
submission_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/sample_submission.csv" # 示例提交文件路径(如果需要加载的话)

# --- 2.3 数据加载与探索 ---
print("--- 2.3 数据加载与探索 ---")
# 读取数据
train_df = pd.read_csv(train_path)
test_df = pd.read_csv(test_path)

print("训练集 前5行数据:")
print(train_df.head())
print("\n测试集 前5行数据:")
print(test_df.head())

print("\n训练集 基本信息:")
# train_df.info() # 输出信息较多，报告中可截取关键部分或截图
print(f"\n训练集 数据形状 (行, 列): {train_df.shape}")
print(f"测试集 数据形状 (行, 列): {test_df.shape}")

print("\n训练集 数值特征统计描述:")
print(train_df.describe())

# 可视化目标变量 SalePrice 的分布 (可选，但推荐)
plt.figure(figsize=(8, 5))
sns.histplot(train_df['SalePrice'], kde=True)
plt.title('房屋售价 (SalePrice) 分布')
plt.xlabel('售价')
plt.ylabel('频数')
plt.grid(True)
plt.show()

# --- 2.4 数据预处理 ---
print("\n--- 2.4 数据预处理 ---")

# 1. 特征选择：仅选择数值型特征 (简化流程)
train_numeric = train_df.select_dtypes(include=[np.number])
test_numeric = test_df.select_dtypes(include=[np.number])

# 分离特征和目标变量
X_all_raw = train_numeric.drop(["SalePrice", "Id"], axis=1)
y_all = train_numeric["SalePrice"] # 目标变量
X_test_raw = test_numeric.drop(["Id"], axis=1)

# 确保测试集和训练集具有相同的特征列 (很重要，以防万一某列只在其中一个存在)
common_cols = list(X_all_raw.columns)
X_test_raw = X_test_raw[common_cols]

print(f"选择数值特征后，训练特征形状: {X_all_raw.shape}")
print(f"选择数值特征后，测试特征形状: {X_test_raw.shape}")

# 2. 缺失值处理：使用中位数填充
# 计算训练集的中位数
median_values = X_all_raw.median()
# 填充训练集和测试集（注意：测试集必须用训练集计算出的中位数填充）
X_all_filled = X_all_raw.fillna(median_values)
X_test_filled = X_test_raw.fillna(median_values)

# 检查填充后是否还有缺失值
print(f"\n填充后训练集缺失值总数: {X_all_filled.isnull().sum().sum()}") # 应为 0
print(f"填充后测试集缺失值总数: {X_test_filled.isnull().sum().sum()}") # 应为 0

# 3. 数据划分：将填充后的数据划分为训练集和验证集
X_train, X_val, y_train, y_val = train_test_split(
    X_all_filled, y_all, test_size=0.2, random_state=42 # random_state保证每次划分一致
)
print(f"\n划分后：训练集特征 {X_train.shape}, 验证集特征 {X_val.shape}")
print(f"划分后：训练集目标 {y_train.shape}, 验证集目标 {y_val.shape}")

# 4. 特征缩放 (标准化)
# 注意：线性回归对特征缩放不是必须的，这里为了匹配你的原始 LR 代码，不进行缩放。
# 如果需要缩放，取消下面的注释并在fit/predict时使用 _scaled 变量。
# print("\n线性回归模型，此处不进行特征缩放 (StandardScaler)。")
# scaler = StandardScaler()
# X_train_scaled = scaler.fit_transform(X_train)
# X_val_scaled = scaler.transform(X_val)

# --- 2.5.1 线性回归模型 ---
print("\n--- 2.5.1 线性回归模型 ---")

# 1. 构建和训练模型
# 使用未缩放的数据训练线性回归模型
lin_model = LinearRegression()
print("开始训练线性回归模型...")
lin_model.fit(X_train, y_train)
print("模型训练完成。")
# print("Linear Regression Coefficients:", lin_model.coef_) # 系数通常很多，报告中可能不需要全部列出
# print("Linear Regression Intercept:", lin_model.intercept_)

# 3. 在验证集上预测
y_val_pred_lin = lin_model.predict(X_val)

# 4. 评估模型
rmse_lin = np.sqrt(mean_squared_error(y_val, y_val_pred_lin))
r2_lin = r2_score(y_val, y_val_pred_lin)
print(f"\n线性回归 - 验证集 RMSE: {rmse_lin:.4f}")
print(f"线性回归 - 验证集 R2: {r2_lin:.4f}")

# 5. 可视化：真实 vs 预测散点图 (验证集)
plt.figure(figsize=(8, 6))
plt.scatter(y_val, y_val_pred_lin, alpha=0.6, color='blue', label='预测值')
plt.plot([y_val.min(), y_val.max()], [y_val.min(), y_val.max()], 'r--', label='理想预测 (y=x)')
plt.xlabel("真实房价 (验证集)")
plt.ylabel("预测房价 (线性回归)")
plt.title("线性回归模型 - 真实值 vs 预测值")
plt.legend()
plt.grid(True)
plt.show()

# 6. 可视化：残差分布图 (验证集)
residuals_lin = y_val - y_val_pred_lin
plt.figure(figsize=(8, 5))
sns.histplot(residuals_lin, bins=30, kde=True)
plt.title("线性回归模型 - 残差分布图")
plt.xlabel("残差 (真实值 - 预测值)")
plt.ylabel("频数")
plt.axvline(0, color='red', linestyle='--', linewidth=1) # 添加均值为0的参考线
plt.grid(True)
plt.show()

# --- 最终预测与提交 (可选) ---
print("\n--- 生成线性回归预测提交文件 ---")
# 使用全部训练数据重新训练最终模型
final_model_lr = LinearRegression()
final_model_lr.fit(X_all_filled, y_all) # 使用填充后的全部训练数据

# 对预处理后的测试集进行预测
predictions_lr = final_model_lr.predict(X_test_filled)

# 生成提交文件
submission_lr = pd.DataFrame({
    "Id": test_df["Id"], # 使用原始测试集的Id列
    "SalePrice": predictions_lr
})
submission_lr.to_csv("submission_lr.csv", index=False)
print("线性回归预测结果已保存为 submission_lr.csv")

print("\n线性回归流程结束。")
